presto-main/src/main/java/com/facebook/presto/sql/gen/JoinCompiler.java

```diff
@@ -405,22 +404,22 @@ private static void generateRowEqualsRowMethod(
         Parameter leftPage = arg("leftPage", Page.class);
         Parameter rightPosition = arg("rightPosition", int.class);
         Parameter rightPage = arg("rightPage", Page.class);
+        Parameter pagesHashChannels = arg("pagesHashChannels", int[].class);
         MethodDefinition rowEqualsRowMethod = classDefinition.declareMethod(
                 a(PUBLIC),
                 "rowEqualsRow",
                 type(boolean.class),
                 leftPosition,
                 leftPage,
                 rightPosition,
-                rightPage);
+                rightPage,
+                pagesHashChannels);
 
-        Scope compilerContext = rowEqualsRowMethod.getScope();
         for (int index = 0; index < joinChannelTypes.size(); index++) {
             BytecodeExpression type = constantType(callSiteBinder, joinChannelTypes.get(index));
 
-            BytecodeExpression leftBlock = leftPage.invoke("getBlock", Block.class, constantInt(index));
-
-            BytecodeExpression rightBlock = rightPage.invoke("getBlock", Block.class, constantInt(index));
+            BytecodeExpression leftBlock = leftPage.invoke("getBlock", Block.class, pagesHashChannels.getElement(index));
+            BytecodeExpression rightBlock = rightPage.invoke("getBlock", Block.class, pagesHashChannels.getElement(index));
```

Comment:
> This is done so the VM can inline the equality code into the generated code, because it always knows the columns are types that are being checked at compile time.

I don't understand what's the difference between old and my code in that regard. They should optimize in the same manner except of one additional memory access to obtain column number, instead of having hardcoded integer for that. 

I'm not so sure whether it would be faster or slower compared to previous version because:
- old version required reordering blocks in page and creating new arrays for columns
- in my version `int[] hashChannels` should be cashed, since it's a small table 
- it is being called only limited number of times - `min(ROWS_IN_PAGE, AVERAGE_GROUP_SIZE)`. Especially for not pre-grouped/pre-sorted pages it will be called only once. For such cases I would expect my version to be faster (old version would rearrange Page just to perform two comparisons...).

Btw @erichwang marked it as a "to do" to fix it as I did it. Maybe we can ask him about this?